### ISSUE TITLE
Polkadot node warning "unexpected canonical chain state…"

### ISSUE THREAD

**Issue by ggwpez** [paritytech/substrate#11262]
I synced a Polkadot from a Polkachu snapshot, which worked fine.  
Then after some time it printed this:  
```pre
Error importing block 0x62fbc8c30cabdab4d0a92234c1556a26c6bc090cf4c14b9eafca6e4ca30311bc: consensus error: Import failed: Parent block of 0x62fb…11bc has no associated weight
```
It kept importing blocks but would print the following warning every now and then:  
```pre
unexpected canonical chain state or malformed finality notification
```
Here the `-lbabe=trace` log with these warnings in them: [log.txt](https://github.com/paritytech/substrate/files/8533320/log.txt).  
@bkchr pobably knows what this means :smiley:

**Comment by bkchr** [comment]
With https://github.com/paritytech/substrate/pull/11263

> unexpected canonical chain state or malformed finality notification

should be gone.

I think the other error may comes from having a fork block in the queue while we already have finalized past this fork and then we get this error?

WDYT @andresilva @davxy

**Comment by andresilva** [comment]
> I think the other error may comes from having a fork block in the queue while we already have finalized past this fork and then we get this error?

I think this can definitely happen. One option would be to add some leeway and prune N blocks after last finalized.

**Comment by bkchr** [comment]
Yeah, we could probably store the finalized heads/stale heads in the aux store and then remove them later when something else gets finalized.
